src/include/platform/internal/GenericPlatformManagerImpl_POSIX.cpp

```diff
@@ -250,29 +294,96 @@ CHIP_ERROR GenericPlatformManagerImpl_POSIX<ImplClass>::_StartEventLoopTask()
     SuccessOrExit(err);
     err = pthread_attr_setschedpolicy(&mChipTaskAttr, SCHED_RR);
     SuccessOrExit(err);
+
+    //
+    // We need to grab the lock here since we have to protect setting
+    // mHasValidChipTask, which will be read right away upon creating the
+    // thread below.
+    //
+    pthread_mutex_lock(&mStateLock);
+
     err = pthread_create(&mChipTask, &mChipTaskAttr, EventLoopTaskMain, this);
-    SuccessOrExit(err);
+    if (err == 0)
+    {
+        mHasValidChipTask = true;
+        mTaskType         = kInternallyManagedTask;
+    }
+
+    pthread_mutex_unlock(&mStateLock);
+
 exit:
     return System::MapErrorPOSIX(err);
 }
 
 template <class ImplClass>
-CHIP_ERROR GenericPlatformManagerImpl_POSIX<ImplClass>::_Shutdown()
+CHIP_ERROR GenericPlatformManagerImpl_POSIX<ImplClass>::_StopEventLoopTask()
 {
     int err = 0;
+
+    //
+    // Signal to the runloop to stop.
+    //
     mShouldRunEventLoop.store(false, std::memory_order_relaxed);
-    if (mChipTask)
+
+    pthread_mutex_lock(&mStateLock);
+
+    //
+    // If we're calling this from a different thread than the one running chip, then
+    // we need to wait till the event queue has completely stopped before proceeding.
+    //
+    if (mHasValidChipTask && (pthread_equal(pthread_self(), mChipTask) == 0))
     {
+        pthread_mutex_unlock(&mStateLock);
+
+        //
+        // We need to grab the lock to protect critical sections accessed by the WakeSelect() call within
+        // SystemLayer.
+        //
+        Impl()->LockChipStack();
         SystemLayer.WakeSelect();
-        SuccessOrExit(err = pthread_join(mChipTask, nullptr));
+        Impl()->UnlockChipStack();
+
+        pthread_mutex_lock(&mStateLock);
+
+        while (!mEventQueueHasStopped)
+        {
+            err = pthread_cond_wait(&mEventQueueStoppedCond, &mStateLock);
+            SuccessOrExit(err);
+        }
+
+        pthread_mutex_unlock(&mStateLock);
+
+        //
+        // Wait further for the thread to terminate if we had previously created it.
+        //
+        if (mTaskType == kInternallyManagedTask)
+        {
+            err = pthread_join(mChipTask, nullptr);
```

Comment:
Despite our earlier conversation, I added this back as an additional step @msandstedt since tsan was complaining that I was leaking a thread every single time, so I figured this is the 'cleanest' thing to do.